### PR TITLE
use a QMimeData clipboard object to paste features as HTML table

### PR DESCRIPTION
**Pull request by nirvn**
## Description
Until now, when pasting copied features into a word processor or a browser's rich text area, the result is a series of non-formatted text lines. It can be frustrating when copying from the attribute table and expecting a table when pasting. The user is left with opening a spreadsheet program, pasting the clipboard content into a sheet first.

This PR fixes that gab by using a QMimeData object and passing in both text as well as HTML content.

Increased productivity ensues:
giphy

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nyalldawson** [comment]
Neat!

I have one reservation here - if I copy a whole ton of points (many thousands), then this is going to dramatically increase the memory required. Could we put a reasonable feature limit here, e.g. only generating the html mime data if there's < 5000 features selected?

Also a unit test ;)

**Comment by nirvn** [comment]
@nyalldawson , comments addressed.

Here's another screencast, showing copy -> pasting into a word document:
giphy

I'll merge this fix/improvement over the weekend if there are no objections.

File: src/app/qgsclipboard.cpp [hunk 1]
```diff
@@ -154,22 +154,38 @@ void QgsClipboard::setSystemClipboard()
...
+  CopyFormat format = AttributesWithWKT;
+  if ( settings.contains( QStringLiteral( "/qgis/copyFeatureFormat" ) ) )
+  {
+    format = static_cast< CopyFormat >( settings.value( QStringLiteral( "qgis/copyFeatureFormat" ), true ).toInt() );
+  }
+
+  if ( format == AttributesWithWKT && mFeatureClipboard.count() < 1000 )
+  {
+    QString htmlCopy = textCopy;
+    htmlCopy.replace( QRegExp( "\n" ), QStringLiteral( "</td></tr><tr><td>" ) );
```
Comment (nyalldawson) [inline src/app/qgsclipboard.cpp]: This doesn't need to be a regexp replace, does it? A character based replace will be MUCH faster

    htmlCopy.replace( '\n', QStringLiteral(".....") );

File: src/app/qgsclipboard.cpp [hunk 2]
```diff
@@ -154,22 +154,38 @@ void QgsClipboard::setSystemClipboard()
...
+  m->setText( textCopy );
+
+  QgsSettings settings;
+  CopyFormat format = AttributesWithWKT;
+  if ( settings.contains( QStringLiteral( "/qgis/copyFeatureFormat" ) ) )
+  {
+    format = static_cast< CopyFormat >( settings.value( QStringLiteral( "qgis/copyFeatureFormat" ), true ).toInt() );
+  }
+
+  if ( format == AttributesWithWKT && mFeatureClipboard.count() < 1000 )
```
Comment (nyalldawson) [inline src/app/qgsclipboard.cpp]: Can you change this to a switch please? (It'll be safer if additional formats are added in future - a warning will be thrown and the developer will need to consider this code block too)

**Comment by nirvn** [comment]
@nyalldawson , comments addressed.

**Pull request merged**